requirements.txt

```
@@ -1,3 +1,5 @@
-lion-pytorch
-pytest
-scipy
+pytest==7.2.2
+einops==0.6.0
+wheel==0.40.0
+lion-pytorch==0.0.6
+scipy==1.11.4
```

Please loosen these up some – chances are an exact version will not necessarily be available, especially not as a wheel, for all platforms + Python versions. [The `~=` operator](https://stackoverflow.com/a/39590286/51685) should be good here. (And let's sort them for good measure...)

```suggestion
einops~=0.6.0
lion-pytorch~=0.0.6
pytest~=7.2.2
scipy~=1.11.4
wheel~=0.40.0
```

---

Note that these are build time dependencies, not run time dependencies. I would actually argue that pinning dependencies hard is a good thing while building. I view the requirements txt file as the lock file other package managers use. This is the only way to get reproducible builds AFAIK. Or is there a better way?

---

These are really test-time dependencies, so pinning them to an exact version may cause any developer's `pip install -r requirements.txt` to fail in creative ways.

Dependencies required to build a [PEP517](https://peps.python.org/pep-0517/) package should be declared according to [PEP518](https://peps.python.org/pep-0518/), i.e. in `pyproject.toml`, and this package already does so :) IOW, `pip install -r requirements.txt` should not be necessary to build this package if it's built with `python -m build .` (as recommended elsewhere in my comments).

We're not guaranteed bit-for-bit reproducible builds anyway since the GHA build infrastructure isn't immutable...

---

Not bit for bit, but isn't it considered good practice when I make a PR to build/test on the exact same version of dependencies as the target branch? I don't know if Python is any different here, but the normal workflow I've grown accustomed to is that a CI system should not arbitrarily upgrade packages from what the developer uses.

And I don't understand why the pip install would fail if we test on all supported platforms? Isn't it the other way around ,that an arbitrary patch upgrade might for some obscure reason break tests in an unrelated PR?

Correct me if I am wrong but I think PEP518 is concerned with the build system itself, not runtime dependencies for running the tests.

Is it considered bad practice in Python to pin versions in requirements.txt? I always learned to pip freeze > requirement.txt when working with CI.

---

>  isn't it considered good practice when I make a PR / is it considered bad practice in Python to pin versions in requirements.txt

IMO, the answer is "it depends". The libraries mentioned here should be prominent and well-behaved enough that they follow semver and minor/patch version increases should not cause issues.

> And I don't understand why the pip install would fail if we test on all supported platforms? 

The issue is that `requirements.txt` here is not only for CI – running tests locally requires these dependencies to be installed, and as time goes by and developers e.g. change platforms, upgrade Pythons, etc., a too strictly specified (then old) version will not have wheels/compatibility with that platform, leading to another kind of cryptic installation failure.

> Correct me if I am wrong but I think PEP518 is concerned with the build system itself

Yep! If the workflow is refactored (as discussed in the other comments, I think) to build a wheel once, these wouldn't be required for the build wheel step. For tests, I'd probably prefer `pip install -e .[test]` or similar, i.e. that the test dependencies are declared in an `extra`.

> I always learned to `pip freeze > requirement.txt` when working with CI.

That's not quite foolproof either. If a package depends on e.g. `pywin32` when running on Windows, this would create a requirements file that will absolutely always fail to install on any non-Windows platform.

---

> should be prominent and well-behaved enough that they follow semver

IMHO this is not enough. Upgrading even patch versions should be a manual step. Hopefully automated, with Dependabot, but still something that will result in a code change. This is what lockfiles are for, and I understand that the concept in Pip currently is not perfect.

> The issue is that requirements.txt here is not only for CI – running tests locally requires these dependencies to be installed, and as time goes by and developers e.g. change platforms

Why not just update the dependencies in this case? Wouldn't you want this to be the new baseline for tests in this case?

I'm happy to change in whichever way the maintainers see suitable but I do want an argument for pinning versions in CI builds.

---

Yes, that does sound like a good idea. This would be more like the way Yarn or Cargo does it.

---

@akx I tried to read up some around the best practices on how to build Python packages today, and made an attempt to move in that direction. You should now have commit rights to this branch, so feel free to fix anything you
think can be done better.

Key points:
* Use the `build` module to build
* Move dev time dependencies to requirements-dev.txt and CI dependencies to requirements-ci.txt (the difference is that the latter have pinned versions). I verified that Dependabot handles these two files as expected. This can be automated better by using something like `poetry` but I think that needs to be discussed a bit first.
* I did not yet move the wheel building, but I agree in principle with your suggestion to build the wheels in one step, I just couldn’t figure out quickly how to do the platform specific wheels in a standard Linux runner. Feel free to fix if you want :)

---

I settled with a "manually" managed requirements-ci.txt and one for local dev with floating versions. These will be handled correctly by dependabot, so as long as they aren't incorrectly updated this will work and they will be kept in sync. For the future, I believe a more advanced build system like Poetry would be useful since it can manage all of this with just one source of truth and one lockfile. But I think this can come in a  subsequent PR

---

I could just skim your conversation, but I generally really like working with lockfiles and explicitly specified dev dependencies.

Conda-lock lockfiles btw are multi-platform now. Anyways, the issue with conda is that the pip only dependencies don't mix well, right? Or am I wrong in that assumption?

I'm all in for implementing modern packaging practices in this repo. Looking forward to that.

---

I think we shouldn't mix anything conda-related in here, simply because we don't need to and it's an entirely different ecosystem. But for a library, the dependencies declared in the package (i.e. not `requirements.txt`) **must** be loosely pinned to avoid conflicts. But we'll get there when we get to looking at packaging...

---

I agree with @akx here, conda is more geared towards setting up dev machines and not so much for CI building of PyPi packages. I would argue we should look at something like [Poetry](https://python-poetry.org/docs/dependency-specification/) if we want something more advanced than what is in this PR. Admittedly I've never worked with Poetry but I skimmed through their documentation and examples and it looks really nice to work with. Anyone with real world experience of Poetry or any high level pip-compatible build backend naturally overrides what I just wrote :)

I would argue we stick with just simple, non-opinonated Python build tools first, then look to migrate in subsequent PRs. Using the `build` package like it does here should serve as a good baseline IMHO.

---

I would recommend Hatch over Poetry (but that's for a future PR).

---

> I would recommend Hatch over Poetry (but that's for a future PR).

Interesting, yes that sounds good too. I just went by the GitHub start count to be honest, but practical experience wins as usual :)